todoapp/TODOapp/app/src/main/java/com/nearsoft/androidschool/todoapp/activities/detail/DetailActivity.java

```diff
@@ -5,50 +5,163 @@
+            }
+        });
+    }
+
+    private void setEditListener() {
         editFab.setOnClickListener(new View.OnClickListener() {
             @Override
+            public void onClick(View view) {
+                buttonViewConfig(true);
+            }
+        });
+    }
+
+    private void setDateListener() {
+        dateCardView.setOnClickListener(new View.OnClickListener() {
+            @Override
             public void onClick(View v) {
-                Snackbar.make(editFab, "should open the \"edit\" activity", Snackbar.LENGTH_SHORT).show();
+                showDatePicker();
             }
         });
+    }
 
-        titleTextView.setText(todo.getTitle());
-        notesTextView.setText(todo.getNotes());
-        dateTextView.setText(todo.getDate());
+    private void showDatePicker() {
+        DatePickerFragment dialog = new DatePickerFragment();
+        dialog.attachListener(new android.app.DatePickerDialog.OnDateSetListener() {
+            @Override
+            public void onDateSet(DatePicker datePicker, int year, int month, int dayOfMonth) {
+                Date dateTime = new Date(year, month + 1, dayOfMonth, 0, 0);
+                updateDate(dateTime);
+            }
+        });
+        dialog.show(this.getSupportFragmentManager(), "datePicker");
     }
 
-    private ToDoContent getTodo() {
+    private void setSwitchListener() {
+        dateSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
+            @Override
+            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
+                dateCardView.setVisibility(isChecked ? View.VISIBLE : View.INVISIBLE);
+            }
+        });
+    }
+
+    private void buttonViewConfig(boolean isEditClicking) {
+        titleEditTextView.setEnabled(isEditClicking);
+        notesEditTextView.setEnabled(isEditClicking);
+        dateCardView.setEnabled(isEditClicking);
+        dateSwitch.setEnabled(isEditClicking);
+        editFab.setVisibility(isEditClicking ? View.INVISIBLE : View.VISIBLE);
+        saveFab.setVisibility(isEditClicking ? View.VISIBLE : View.INVISIBLE);
+    }
+
+    private void saveData() throws AssertionError {
+//        TODO:this method will make the saving
+        String selectedDate = dateTextView.getText().toString();
+        if (selectedDate.equals(getText(R.string.date))) {
+            throw new AssertionError(dateCardView.getClass().getSimpleName() + getString(R.string.set_date_or_change_switch_message));
+        }
+        todoItem.setTitle(titleEditTextView.getText().toString());
+        todoItem.setDate(new Date(selectedDate));
+        todoItem.setNotes(notesEditTextView.getText().toString());
+        todoItem.setHasDate(dateSwitch.isChecked());
+    }
+
+    private void populateToDoObject() {
+        try {
+            todoItem = getTodo();
+            displayDetail();
+            buttonViewConfig(false);
+        } catch (AssertionError err) {
```

Comment:
if the problem is this line:
`throw new AssertionError(getClass().getSimpleName() + " intent extras should contain an item");`

and it is totally possible now to create a `DetailActivity.class` without extras, just create a new Todo there, instead of continue throwing the AssertionError :D